cmd/frontend/graphqlbackend/search.go

```
@@ -342,24 +342,40 @@ func decodedViewerFinalSettings(ctx context.Context) (*schema.Settings, error) {
 	return &settings, nil
 }
 
-var mockResolveRepoGroups func() (map[string][]*types.Repo, error)
+var mockResolveRepoGroups func() (map[string][]*types.Repo, []string, error)
 
-func resolveRepoGroups(settings *schema.Settings) (map[string][]*types.Repo, error) {
+type regexPath struct {
+	name  string
+	regex string
+}
+
+func resolveRepoGroups(settings *schema.Settings) (groups map[string][]*types.Repo, patterns []string, err error) {
 	if mockResolveRepoGroups != nil {
 		return mockResolveRepoGroups()
 	}
-
-	groups := map[string][]*types.Repo{}
-
-	for name, repoPaths := range settings.SearchRepositoryGroups {
-		repos := make([]*types.Repo, len(repoPaths))
-		for i, repoPath := range repoPaths {
-			repos[i] = &types.Repo{Name: api.RepoName(repoPath)}
+	groups = map[string][]*types.Repo{}
+
+	for name, repoSpecs := range settings.SearchRepositoryGroups {
+		repos := make([]*types.Repo, 0, len(repoSpecs))
+
+		for _, repoSpec := range repoSpecs {
+			switch path := repoSpec.(type) {
+			case string:
+				patterns = append(patterns, "^"+regexp.QuoteMeta(path)+"$")
+				repos = append(repos, &types.Repo{Name: api.RepoName(path)})
+			case map[string]interface{}:
+				if stringRegex, ok := path["regex"].(string); ok {
+					patterns = append(patterns, stringRegex)
+				} else {
+					log15.Warn("ignoring repo group object because regex not specfied", "regex-string", path["regex"])
+				}
+			default:
+				log15.Warn("ignoring repo group object of unrecognized type", "object", repoSpec, "type", fmt.Sprintf("%T", repoSpec))
```

I think we should return an error? This would indicate us breaking sourcegraph and a log message will be ignored. if we are worried about someone missing this in the future, maybe this parsing of configuration can be pulled out and tested?

---

It feels like a warning is more appropriate here. Otherwise, a mistake in the `search.repositoryGroups` field of global settings will break all searches. Currently, the editor will display yellow warning squiggles, but it won't prevent you from saving the bad config that has invalid entries in `search.repositoryGroups`.